From: Balint Reczey <balint.reczey@canonical.com>
Date: Wed, 24 Apr 2019 17:24:02 +0200
Subject: Add check to switch VTs only between K_XLATE or K_UNICODE

Switching to K_UNICODE from other than L_XLATE can make the keyboard
unusable and possibly leak keypresses from X.

BugLink: https://launchpad.net/bugs/1803993
(cherry picked from commit 13a43c73d8cbac4b65472de04bb88ea1bacdeb89)
---
 src/basic/terminal-util.c     | 9 ++++++++-
 src/vconsole/vconsole-setup.c | 7 +++++++
 2 files changed, 15 insertions(+), 1 deletion(-)

diff --git a/src/basic/terminal-util.c b/src/basic/terminal-util.c
index 48ede7d..c7a7455 100644
--- a/src/basic/terminal-util.c
+++ b/src/basic/terminal-util.c
@@ -1273,11 +1273,18 @@ int vt_verify_kbmode(int fd) {
 }
 
 int vt_reset_keyboard(int fd) {
-        int kb;
+        int kb, r;
 
         /* If we can't read the default, then default to unicode. It's 2017 after all. */
         kb = vt_default_utf8() != 0 ? K_UNICODE : K_XLATE;
 
+        r = vt_verify_kbmode(fd);
+        if (r == -EBUSY) {
+                log_debug_errno(r, "Keyboard is not in XLATE or UNICODE mode, not resetting: %m");
+                return 0;
+        } else if (r < 0)
+                return r;
+
         if (ioctl(fd, KDSKBMODE, kb) < 0)
                 return -errno;
 
diff --git a/src/vconsole/vconsole-setup.c b/src/vconsole/vconsole-setup.c
index be09619..bfa0f57 100644
--- a/src/vconsole/vconsole-setup.c
+++ b/src/vconsole/vconsole-setup.c
@@ -74,6 +74,13 @@ static int toggle_utf8(const char *name, int fd, bool utf8) {
 
         assert(name);
 
+        r = vt_verify_kbmode(fd);
+        if (r == -EBUSY) {
+                log_warning_errno(r, "Virtual console %s is not in K_XLATE or K_UNICODE: %m", name);
+                return 0;
+        } else if (r < 0)
+                return log_warning_errno(r, "Failed to verify kbdmode on %s: %m", name);
+
         r = ioctl(fd, KDSKBMODE, utf8 ? K_UNICODE : K_XLATE);
         if (r < 0)
                 return log_warning_errno(errno, "Failed to %s UTF-8 kbdmode on %s: %m", enable_disable(utf8), name);
